src/core/core.c

```diff
@@ -88,6 +88,11 @@ static void sig_reload_config(int signo)
         reload_config = TRUE;
 }
 
+static void sig_quit_on_hup(int signo)
+{
+	signal_emit("gui exit", 0);
```

Comment:
This isn't [async-signal safe][1]. The code can safely inside signal handlers is very limited, and there's stuff inside signal_emit and the gui exit handlers that interacts with the memory allocator, which holds a lock. So if a SIGHUP arrives at the precise moment some other part of the code is doing memory allocation stuff anywhere else, that'll deadlock irssi.

You have to do something similar to the handler above, which just changes the value of the global `reload_config` to TRUE, and nothing else, just letting some code outside the signal handler do the actual config reload.

Things you could do here:

1. Use the `quitting` global variable, defined at `src/fe-text/irssi.c`. This will quit cleanly, but it means other signal handlers of `gui exit` won't run, like autosave.
2. Drop this separate handler, and repurpose/rename the `reload_config` flag to a generic `sighup_received`, and adapt the current `reload_config` handler implemented in `src/fe-text/irssi.c` so that it either reloads the config or quits depending on what the setting says.

[1]: https://wiki.sei.cmu.edu/confluence/display/c/SIG30-C.+Call+only+asynchronous-safe+functions+within+signal+handlers

Reply:
Thanks for the info! I actually didn't know that, as I never used signals before.